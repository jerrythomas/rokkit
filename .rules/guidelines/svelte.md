# Guidelines

- Use Svelte 5 runes mode for code
- Use SvelteKit for app
- Use UnoCSS for styling
- Use vitest for unit testing
- Use Playwright for e2e testing
- E2E tests should follow the page object pattern

## Svelte 5
{{[Svelte & Svelte Kit Distilled](https://svelte-llm.khromov.se/svelte-complete-distilled)}}
{{[svelte 5 documentation](https://svelte.dev/docs/svelte/llms.txt)}}

## Snippets

### Runes

Runes are special functions that replace reactivity declarations from Svelte 4.

```svelte
<script>
  // State
  let count = $state(0);

  // Derived values
  let doubled = $derived(count * 2);

  // Props
  let {
    name = 'world',
    value = $bindable(null)
  } = $props();

  // Effects
  $effect(() => {
    console.log(`Count changed to ${count}`);
  });
</script>
```

### Defining and passing snippets to components:

```svelte
<script>
  let items = ['one','two', 'three']
</script>
<!-- a header snippet -->
{#snippet header()}
  <h1>Custom header</h1>
{/snippet}

<!-- a snippet with parameters -->
{#snippet item(item, selected)}
  <div class:selected={selected}>
    {item.name}
  </div>
{/snippet}

<List {items} {header} {item} />
```

### Using snippets in a component:

```svelte
<script>
  let { header, item } = $props();
</script>

<div class="header">
  {@render header?.()}
</div>

{#each items as currentItem}
  {@render item(currentItem, selectedItem === currentItem)}
{/each}
```

To pass a snippet to another component:

#### ChildComponent.svelte

```svelte
<script>
  let { header } = $props();
</script>

{#if header}
  {@render header()}
{/if}
```

#### Snippets can be passed as properties to child component

```svelte
<script>
   import  ChildComponent from './ChildComponent.svelte'
</script>
{#snippet header()}
  <h1>Custom Header</h1>
{/snippet}

<ChildComponent {header} />
```

### Snippets can be enclosed within the child component.

```svelte
<script>
   import  ChildComponent from './ChildComponent.svelte'
</script>
<ChildComponent>
  {#snippet header()}
    <h1>Custom Header</h1>
  {/snippet}
</ChildComponent>
```
Both of these options work in a similar way

### Important: What NOT to do

Do not confuse snippets with React-style JSX or string templates:

```svelte
<!-- ❌ INCORRECT - Using arrow functions returning strings -->
<List
  header={() => `<h1>My List</h1>`}
  item={({ item }) => `<div>${item.name}</div>`}
/>

<!-- ✅ CORRECT - Using properly defined snippets -->
<!-- snippets are defined in the html section not in scripts block -->
{#snippet header()}
  <h1>My List</h1>
{/snippet}

{#snippet itemTemplate(item, selected)}
  <div class:selected={selected}>
    {item.name}
  </div>
{/snippet}

<List {header} item={itemTemplate} />
```

## Component Design Guidelines

- should be modular and data driven
- should be accessibile
- Reduced motion should be considered

## Project Structure

If the source code uses runes or the test uses runes the file extension should be `.svelte.js` or `.spec.svelte.js` respectively.

```
project/
│
├── src/
│   ├── index.js
│   ├── Component.svelte
│   └── utils.svelte.js
│
├── spec/
│   ├── index.spec.js
│   ├── Component.spec.js
│   └── utils.spec.svelte.js
├── e2e/
│   ├── home.e2e.js
│   └── login.e2e.js
└── README.md
```
