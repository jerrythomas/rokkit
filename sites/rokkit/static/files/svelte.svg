<svg width="24" height="24" viewBox="0 0 24 24" fill="none" xmlns="http://www.w3.org/2000/svg">
<path d="M18.2301 11.1217C21.1087 6.47773 15.3383 1.6935 11.2372 4.37297L7.337 6.85957C5.44834 8.08961 4.62832 10.453 5.78004 12.8817C2.90038 17.2501 8.50365 22.3655 12.7729 19.6332L16.6732 17.1466C18.7952 15.7494 19.2821 13.3398 18.2301 11.1217V11.1217ZM8.46536 8.46761L12.3599 5.98242C14.9184 4.45077 17.8644 7.12896 16.8023 9.62435C16.1715 9.12318 15.4685 8.81503 14.9345 8.67627C15.1375 7.87006 14.2016 7.27202 13.5507 7.68841L9.64763 10.1651C8.68258 10.8321 9.71148 12.204 10.6653 11.6284L12.1527 10.6803C15.4905 8.79676 18.6495 13.3192 15.5448 15.5357L11.6502 18.0209C9.30336 19.5215 6.10922 17.1175 7.20785 14.379C7.79145 14.8207 8.41131 15.1456 9.07564 15.3271C8.92896 16.2531 9.86333 16.7033 10.4694 16.3234L14.3626 13.8382C15.3233 13.1178 14.292 11.814 13.3464 12.375L11.8533 13.3231C8.55362 15.2514 5.32106 10.6843 8.46536 8.46761Z" fill="black"/>
</svg>
