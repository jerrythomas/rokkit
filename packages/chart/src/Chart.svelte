<script>
	import { run } from 'svelte/legacy';

	import { setContext } from 'svelte'
	import { writable } from 'svelte/store'
	import { chart } from './lib'

	let config = writable({})
	setContext('chart', config)

	
	/**
	 * @typedef {Object} Props
	 * @property {any} data
	 * @property {any} x
	 * @property {any} y
	 * @property {any} [value]
	 * @property {any} [color]
	 * @property {any} [fill]
	 * @property {number} [width] - export let pattern = x
	 * @property {number} [height]
	 * @property {any} [padding]
	 * @property {number} [marginLeft]
	 * @property {number} [marginRight]
	 * @property {number} [marginTop]
	 * @property {number} [marginBottom]
	 * @property {boolean} [flipCoords]
	 * @property {number} [spacing]
	 * @property {import('svelte').Snippet} [children]
	 */

	/** @type {Props} */
	let {
		data,
		x,
		y,
		value = y,
		color = x,
		fill = x,
		width = 2048,
		height = 2048,
		padding = height / 16,
		marginLeft = 0,
		marginRight = 0,
		marginTop = 0,
		marginBottom = 0,
		flipCoords = false,
		spacing = 0.1,
		children
	} = $props();

	let margin = $derived({
		left: marginLeft,
		right: marginRight,
		top: marginTop,
		bottom: marginBottom
	})
	// $: patterns = [...new Set(data.map((d) => d[pattern]))]
	// $: fills = [...new Set(data.map((d) => d[fill]))]
	// $: colors = [...new Set(data.map((d) => d[color]))]
	run(() => {
		config.set(
			chart(data, {
				x,
				y,
				value,
				color,
				fill,
				width,
				height,
				padding,
				margin,
				flipCoords,
				spacing
			})
		)
	});
</script>

<svg
	viewBox="0 0 {$config.width} {$config.height}"
	width={$config.width}
	height={$config.height}
	class="w-full h-full chart"
>
	<rect
		x="0"
		y="0"
		width={$config.width}
		height={$config.height}
		fill="none"
		stroke="currentColor"
	/>
	{@render children?.()}
</svg>
