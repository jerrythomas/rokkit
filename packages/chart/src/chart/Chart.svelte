<script>
	import { setContext } from 'svelte'
	import { writable } from 'svelte/store'
	import { chart } from '../lib/chart'

	let config = writable({})
	setContext('chart', config)

	export let data
	export let x
	export let y
	export let value = y
	export let color = x
	export let fill = x
	export let width = 512
	export let height = 512
	export let padding = height / 16
	export let marginLeft = 0
	export let marginRight = 0
	export let marginTop = 0
	export let marginBottom = 0
	export let flipCoords = false
	export let spacing = 0.1

	$: margin = {
		left: marginLeft,
		right: marginRight,
		top: marginTop,
		bottom: marginBottom
	}
	$: config.set(
		chart(data, {
			x,
			y,
			value,
			color,
			fill,
			width,
			height,
			padding,
			margin,
			flipCoords,
			spacing
		})
	)
</script>

<svg
	viewBox="0 0 {$config.width} {$config.height}"
	width={$config.width}
	height={$config.height}
	class="w-full h-full chart"
>
	<!-- <rect
		x="0"
		y="0"
		width={$config.width}
		height={$config.height}
		fill="none"
		stroke="currentColor"
	/> -->
	<slot />
</svg>
