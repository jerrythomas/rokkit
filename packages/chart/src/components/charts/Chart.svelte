<script>
	import { setContext } from 'svelte'
	import { writable } from 'svelte/store'
	import { compact } from '../lib/utils'
	import { builtIn } from '../lib/theme'

	let chart = writable({})

	setContext('chart', chart)

	export let width = 800
	export let height = 450
	export let data
	export let theme = builtIn
	export let x
	export let y
	export let fill
	export let color
	export let padding = 20
	export let curve = 'basis'
	export let stat = 'identity'

	$: aes = compact({ x, y, fill, color, stat, curve, padding })
	$: chart.set({
		width,
		height,
		data,
		theme,
		aes,
	})
</script>

<svg viewBox="0 0 {width} {height}" class="chart" width="100%">
	<slot />
</svg>
