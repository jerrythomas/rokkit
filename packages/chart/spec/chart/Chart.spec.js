import { describe, expect, beforeEach, it } from 'vitest'
import { cleanup, render } from '@testing-library/svelte'
import Chart from '../../src/chart/Chart.svelte'

describe('Chart.svelte', () => {
	beforeEach(() => cleanup())

	it('should render', () => {
		// const { container } = render(Chart)
		// expect(container).toMatchSnapshot()
	})
})
