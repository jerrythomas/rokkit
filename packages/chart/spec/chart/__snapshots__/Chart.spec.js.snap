// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Chart.svelte > should render 1`] = `
<body>
  <div>
    <svg
      class="chart"
      viewBox="0 0 800 450"
      width="100%"
    />
  </div>
</body>
`;
